File: client/java/src/krpc/client/Event.java
package krpc.client;

import krpc.client.Stream;
import krpc.schema.KRPC;
import krpc.schema.KRPC.Error;
import krpc.schema.KRPC.ProcedureCall;
import krpc.schema.KRPC.Type;
import krpc.schema.KRPC.Type.TypeCode;

import java.io.IOException;
import java.lang.Runnable;

public class Event implements Comparable<Event> {
  private Connection connection;
  private Stream<Boolean> stream;

  Event(Connection connection, KRPC.Event event) {
    Type returnType = Type.newBuilder().setCode(TypeCode.BOOL).build();
    stream = new Stream<Boolean>(connection, returnType, event.getStream().getId());
  }

  @Override
    public int hashCode() {
    return stream.hashCode();
  }

  @Override
  @SuppressWarnings("unchecked")
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null) {
      return false;
    }
    if (getClass() != obj.getClass()) {
      return false;
    }
    final Event other = (Event) obj;
    return this.stream.equals(other.stream);
  }

  @Override
    public int compareTo(final Event obj) {
    return this.stream.compareTo(obj.stream);
  }

  /**
   * Start the event.
   */
  public void start() throws RPCException {
    stream.start();
  }

  /**
   * Condition variable that is notified when the event occurs.
   */
  public Object getCondition() {
    return stream.getCondition();
  }

  /**
   * Wait until the event occurs.
   */
  public void waitFor() throws RPCException, StreamException {
    if (!stream.stream.getStarted()) {
      start();
    }
    stream.stream.setValue(false);
    while (!stream.get()) {
      stream.waitForUpdate();
    }
  }

  /**
   * Wait until the event occurs, with a timeout in seconds.
   */
  public void waitForWithTimeout(double timeout) throws RPCException, StreamException {
    if (!stream.stream.getStarted()) {
      start();
    }
    stream.stream.setValue(false);
    while (!stream.get()) {
      boolean origValue = stream.get();
      stream.waitForUpdateWithTimeout(timeout);
      if (stream.get() == origValue) {
        // Value did not change, must have timed out
        return;
      }
    }
  }

  /**
   * Add a callback that is invoked whenever the stream is updated.
   */
  public void addCallback(Runnable callback) {
    stream.addCallback(
        (Boolean value) -> {
          if (value) {
            callback.run();
          }
        });
  }

  /**
   * The underlying stream for the event.
   */
  public Stream<Boolean> getStream() {
    return stream;
  }

  /**
   * Remove the event from the server.
   */
  public void remove() throws RPCException {
    stream.remove();
  }
}

File: client/python/krpc/event.py
from __future__ import annotations
from typing import Callable, Optional, TYPE_CHECKING
import threading
from krpc.stream import Stream
import krpc.schema.KRPC_pb2 as KRPC
if TYPE_CHECKING:
    from krpc.client import Client


class Event:
    """ An event. """
    def __init__(self, client: Client, event: KRPC.Event) -> None:
        self._stream = Stream.from_stream_id(
            client, event.stream.id, client._types.bool_type)
        self._client = client
        self._callback_mapping: \
            dict[Callable[[], None], Callable[[bool], None]] = {}

    def start(self) -> None:
        """ Start the underlying stream for the event """
        self._stream.start(False)

    @property
    def condition(self) -> threading.Condition:
        """ The condition variable for the event """
        return self._stream.condition

    def wait(self, timeout: Optional[float] = None) -> None:
        """ Blocks until the event is triggered or a timeout occurs.
            The condition variable must be locked before calling this method.

            When timeout is not None, it should be a floating point number
            specifying the timeout in seconds for the operation. """
        if not self.stream._stream.started:
            self.start()
        self.stream._stream.value = False
        while not self.stream():
            orig_value = self.stream()
            self.stream.wait(timeout=timeout)
            if timeout is not None and self.stream() == orig_value:
                # Value did not change, must have timed out
                return

    def add_callback(self, callback: Callable[[], None]) -> None:
        def callback_wrapper(x: bool) -> None:
            if x:
                callback()
        self._callback_mapping[callback] = callback_wrapper
        self._stream.add_callback(callback_wrapper)

    def remove_callback(self, callback: Callable[[], None]) -> None:
        if callback in self._callback_mapping:
            callback_wrapper = self._callback_mapping[callback]
            del self._callback_mapping[callback]
            self._stream.remove_callback(callback_wrapper)

    @property
    def stream(self) -> Stream:
        """ The underlying stream for the event """
        return self._stream

    def remove(self) -> None:
        """ Remove the event from the server """
        self._stream.remove()
